### PR TITLE
Feature: Album duration

### LINKED ISSUE #588: [Feature Request] Total duration of an album
### Description

Total duration of the album displayed as a small text at the bottom of the album view.

### Checklist
-   [x] I have read the [Contribution Guidelines](https://github.com/zyrouge/symphony/wiki/Contributions-Guidelines#issues).

### PR DESCRIPTION
**Pull request by sevonj**
Closes #588 
 
## Changes
- Albums now track total duration
- The duration is shown in album view

<details>
<summary> screenshot </summary>

image

</details>
 
## Comments
**Possible further improvements:**  
- I think the UI would look nicer if the menu moved to the top right corner.  
  <details>
  <summary> mockup </summary>

  image
  </details>
- Custom formatting; don't show seconds if duration >= 1 min 

**Alternatives:**
Display the duration in the top bar along with song count.

### CODE REVIEW

**Comment by Currrupted** [comment]
I think both variants look neat, though I'd expect the menu to have that design/position everywhere.

**Review by zyrouge — CHANGES_REQUESTED** [review]

File: app/src/main/java/io/github/zyrouge/symphony/ui/components/GenericGrooveBanner.kt [hunk 1]
```diff
@@ -65,7 +65,7 @@ fun GenericGrooveBanner(
         ) {
             Box(
                 modifier = Modifier
-                    .padding(defaultHorizontalPadding, 32.dp, defaultHorizontalPadding, 12.dp)
+                    .padding(defaultHorizontalPadding, 32.dp, 0.dp, 12.dp)
```
[the remark this answers is not in the document]
Comment (sevonj) [reply app/src/main/java/io/github/zyrouge/symphony/ui/components/GenericGrooveBanner.kt:68]: image

It felt like too much.

**Comment by sevonj** [comment]
I tested the ui with all kinds of strings, and got more consistent results with a two-column layout instead of fiddling with individual row items.

Screenshot From 2024-11-23 19-33-20

**Comment by zyrouge** [comment]
I think the design would be much better if we moved the duration to the left. The text near the more options icons looks uncomfortable since they are not aligned.

File: app/src/main/java/io/github/zyrouge/symphony/ui/view/Album.kt [hunk 2]
```diff
@@ -129,17 +128,21 @@ private fun AlbumHero(context: ViewContext, album: Album) {
             )
         },
         content = {
-            Column {
-                Text(album.name)
-                Row(modifier = Modifier.fillMaxWidth()) {
+            Row {
+                Column(
```
Comment (zyrouge) [inline app/src/main/java/io/github/zyrouge/symphony/ui/view/Album.kt:132]: Revert this and add duration as a row.

**Review by zyrouge — CHANGES_REQUESTED** [review]

**Comment by sevonj** [comment]
> I think the design would be much better if we moved the duration to the left. The text near the more options icons looks uncomfortable since they are not aligned.

Screenshot From 2024-11-23 20-03-31

Like a row?

**Comment by sevonj** [comment]
Maybe the duration could be slightly less bright colored to make it less awkward.

**Comment by zyrouge** [comment]
> > I think the design would be much better if we moved the duration to the left. The text near the more options icons looks uncomfortable since they are not aligned.
> 
> Screenshot From 2024-11-23 20-03-31
> 
> Like a row?

Yup, that looks for now. We could make changes to the UI in the future.

**Comment by zyrouge** [comment]
> Maybe the duration could be slightly less bright colored to make it less awkward.

I was thinking the same but we might add more metadata such as album start/end year which would make it look less awkward.

**Comment by sevonj** [comment]
With the current design, I'd add the metadata to the same row as in

> 2024-2025 | 2h 49m

**Comment by zyrouge** [comment]
> With the current design, I'd add the metadata to the same row as in
> 
> > 2024-2025 | 2h 49m

Yeah, that is how I meant. Don't add them in this PR yet.

**Comment by sevonj** [comment]
This should be OK to merge now.

**Review by zyrouge — APPROVED** [review]

**Comment by zyrouge** [comment]
LGTM!

**Comment by Currrupted** [comment]
I like the way the duration was implemented. However, I have noticed that the album duration includes milliseconds and thus isn't that intuitive.  The duration isn't 45 minutes and 32 seconds (as an example), but 45m 32.124s. Wouldn't it be better to round the duration to the full integer? Alternatively, maybe the digits after the decimal point could be hidden, somehow?

**Comment by sevonj** [comment]
> I like the way the duration was implemented. However, I have noticed that the album duration includes milliseconds and thus isn't that intuitive. The duration isn't 45 minutes and 32 seconds (as an example), but 45m 32.124s. It doesn't really make sense, because normally the length of a song isn't expressed in milliseconds. Wouldn't it be better to round the duration to the full integer? Alternatively, maybe the digits after the decimal point could be hidden, somehow? @sevonj

I mentioned in the top PR comment that the formatting could be improved. If the default also shows milliseconds in some cases, a custom formatter is more necessary than I thought.

**Comment by Currrupted** [comment]
> > I like the way the duration was implemented. However, I have noticed that the album duration includes milliseconds and thus isn't that intuitive. The duration isn't 45 minutes and 32 seconds (as an example), but 45m 32.124s. It doesn't really make sense, because normally the length of a song isn't expressed in milliseconds. Wouldn't it be better to round the duration to the full integer? Alternatively, maybe the digits after the decimal point could be hidden, somehow? @sevonj
> 
> I mentioned in the top PR comment that the formatting could be improved. If the default also shows milliseconds in some cases, a custom formatter is more necessary than I thought.

That makes sense! I must have missed that. For reference, this is what it looks like for me:
signal-2025-01-04-17-03-57-680

I don't know if there is something wrong with my files or the codec, if the duration is so weird. But the format is like this everywhere for me.

**Comment by zyrouge** [comment]
> > > I like the way the duration was implemented. However, I have noticed that the album duration includes milliseconds and thus isn't that intuitive. The duration isn't 45 minutes and 32 seconds (as an example), but 45m 32.124s. It doesn't really make sense, because normally the length of a song isn't expressed in milliseconds. Wouldn't it be better to round the duration to the full integer? Alternatively, maybe the digits after the decimal point could be hidden, somehow? @sevonj
> > 
> > 
> > I mentioned in the top PR comment that the formatting could be improved. If the default also shows milliseconds in some cases, a custom formatter is more necessary than I thought.
> 
> That makes sense! I must have missed that. For reference, this is what it looks like for me: signal-2025-01-04-17-03-57-680
> 
> I don't know if there is something wrong with my files or the codec, if the duration is so weird. But the format is like this everywhere for me.

I've did not have that issue either. I guess the simplest fix would be to convert them to milliseconds to seconds before displaying.

**Pull request merged**